### ISSUE TITLE
Cursor handling in completionItems

### ISSUE THREAD

**Issue by fbricon** [Microsoft/vscode#24683]
Working on https://github.com/eclipse/eclipse.jdt.ls/pull/206, I noticed some inconsistencies in cursor handling when completionItems are returned as Text vs SnippetString

- VSCode Version: 1.11.1
- OS Version: 10.12.3

Steps to Reproduce:

1. When `@Override\npublic void run() {\n\t{{}}\n}` is returned  as Text
``` Received response 'completionItem/resolve - (336)' in 9ms.
Result: {
    "label": "run() : void",
    "kind": 3,
    "detail": "Override method in 'Runnable'",
    "documentation": "When an object implementing interface `Runnable` is used to create a thread, starting the thread causes the object's `run` method to be called in that separately executing thread.\n\nThe general contract of the method `run` is that it may take any action whatsoever.\n\n *  **See Also:**\n    \n     *  java.lang.Thread.run()",
    "sortText": "zzzzzziied",
    "insertTextFormat": 1,
    "textEdit": {
        "range": {
            "start": {
                "line": 5,
                "character": 1
            },
            "end": {
                "line": 5,
                "character": 2
            }
        },
        "newText": "@Override\npublic void run() {\n\t{{}}\n}"
    }
}
``` 
the above completionItem is rendered as :
![](https://content.screencast.com/users/fbricon/folders/Jing/media/967d02c6-3d41-4fb4-9dd9-c63bd7eb7a3e/00000004.png)

2. Now when returning `@Override\npublic String get(String t) {\n\t{{}}\n}` as SnippetString 
```
[Trace - 9:13:20 AM] Received response 'completionItem/resolve - (352)' in 8ms.
Result: {
    "label": "get(String t) : String",
    "kind": 3,
    "detail": "Override method in 'IFace'",
    "sortText": "zzzzzziied",
    "insertTextFormat": 2,
    "textEdit": {
        "range": {
            "start": {
                "line": 4,
                "character": 1
            },
            "end": {
                "line": 4,
                "character": 2
            }
        },
        "newText": "@Override\npublic String get(String t) {\n\t{{}}\n}"
    }
}
```
the `{{}}` is not rendered as cursor position
![](https://content.screencast.com/users/fbricon/folders/Jing/media/1b499018-2a5d-4787-aeca-164a040e160a/00000005.png)

I understand SnippetString is meant as a way to provide richer code completion with tabstops. I guess I could use ${0} instead, but why not supporting {{}} too? Why are there 2 syntaxes to provide cursor position? 

cc @gorkem

**Comment by jrieken** [comment]
>  I guess I could use ${0} instead, but why not supporting {{}} too? Why are there 2 syntaxes to provide cursor position?

The `{{}}`-syntax was an old and undocumented syntax that has leaked into some extensions. We have deprecated this syntax end of last year and replaced it with an official and documented Textmate-style syntax. In May (or June) we will remove support for the old syntax, SnippetString only supports the new syntax. 

See https://github.com/Microsoft/vscode/issues/15573, https://github.com/redhat-developer/vscode-java/issues/99, and https://code.visualstudio.com/docs/editor/userdefinedsnippets#_snippet-syntax

So in short, use `SnippetString` with the official syntax. Thanks and Happy Coding

**Comment by fbricon** [comment]
@jrieken so there's no way to provide cursor position when using plain text?

**Comment by jrieken** [comment]
Sure, via the SnippetString. Unsure why you don't consider that plain text?

**Comment by fbricon** [comment]
I meant plain text as opposed to snippetstring. Not all clients support snippetstrings

**Comment by jrieken** [comment]
Understood, @dbaeumer might know how well `SnippetString` has been adopted

**Comment by dbaeumer** [comment]
SnippetString has been fully adopted in LSP.
